topdown/eval.go

```diff
@@ -2892,6 +2910,172 @@ func (e evalTerm) save(iter unifyIterator) error {
+		return e.save(iter)
+	}
+
+	child := e.e.closure(e.generator)
+	all := true // all generator evaluations yield one successful body evaluation
+	err := child.Run(func(child *eval) error {
+		if !all {
+			// NOTE(sr): Is this good enough? We don't have a "fail EE".
+			// This would do extra work, like iterating needlessly if domain was a large array.
+			return nil
+		}
+		body := child.closure(e.body)
+		body.findOne = true
+		done := false
+		err := body.eval(func(*eval) error {
+			done = true
+			return nil
+		})
+		if !done {
+			all = false
+		}
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	if all {
+		return iter()
+	}
+	return nil
+}
+
+func (e *evalEvery) save(iter unifyIterator) error {
+	// TODO(sr): check traces
+	kvVars := e.expr.Terms.(*ast.Every).KeyValueVars()
+	vars := make([]*ast.Term, 0, len(kvVars))
+	for _, v := range kvVars.Sorted() {
+		vars = append(vars, ast.NewTerm(v))
+	}
+
+	if !e.e.inliningControl.shallow && !e.e.unknown(e.generator, e.e.bindings) {
+		return e.saveUnrolled(vars, kvVars, iter)
+	}
+
+	return e.saveEveryExpr(vars, kvVars, iter)
+}
+
+func (e *evalEvery) saveUnrolled(vars []*ast.Term, kvVars ast.VarSet, iter unifyIterator) error {
+
+	child := e.e.closure(e.generator)
+	e.e.saveSet.Push(vars, child.bindings)
+	e.e.saveStack.PushQuery(nil)
+
+	all := true
+	err := child.Run(func(child *eval) error {
+		if !all {
+			return nil
+		}
+
+		body := child.closure(e.body)
+		body.findOne = true
+		done := false
+		err := body.eval(func(child *eval) error {
+			current := e.e.saveStack.PopQuery()
+			plugged := current.Plug(child.bindings)
+
+			// Type-checking failure means the rule body will never succeed.
+			if !e.e.compiler.PassesTypeCheck(plugged) {
+				return nil // TODO abort?
+			}
+			cp := copypropagation.New(kvVars).WithCompiler(e.e.compiler)
+			plugged = applyCopyPropagation(cp, e.e.instr, plugged)
+
+			for _, expr := range plugged {
+				// Filter __local0__, __local1__ unifications for key/val pairs, since
+				// copy prop will have inlined those already
+				if expr.IsEquality() {
+					v, ok := expr.Operand(0).Value.(ast.Var)
+					if ok && kvVars.Contains(v) {
+						continue
+					}
+				}
+				e.e.saveStack.Push(expr, child.bindings, child.bindings) // bindings?
+			}
+
+			done = true
+			e.e.saveStack.PushQuery(current)
+			return nil
+		})
+		if !done {
+			all = false
+		}
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	e.e.saveSet.Pop()
+	if all {
+		for _, expr := range e.e.saveStack.PopQuery().Plug(e.e.bindings) {
+			if err := e.e.saveExpr(expr, e.e.bindings, iter); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+
+}
+
+func (e *evalEvery) saveEveryExpr(vars []*ast.Term, kvVars ast.VarSet, iter unifyIterator) error {
+
+	child := e.e.child(e.body)
+	child.traceEnter(e.expr) // ?
+
+	var savedQueries ast.Body
+	e.e.saveStack.PushQuery(nil)
+	e.e.saveSet.Push(vars, child.bindings)
+
+	err := child.eval(func(child *eval) error {
```

Comment:
Hrmmm. I'm not sure about this. For example, given the following query:

```rego
every x in p {
  some y in q
  y > x
}
```

This should say "for every x in p there is some y in q such that y is greater than x".

If `p` is unknown and `q` is known this `eval()` will generate multiple outputs. For example if `q` is `[6,7,8]` then this would generate:

```
# query 1
6 > x

# query 2
7 > x

# query 3
8 > x
```

However, we cannot represent that inside of the `ast.Every` expr because the body is a single query.